ext/pcre/php_pcre.c

```diff
@@ -2156,85 +2147,81 @@ static zend_string *php_pcre_replace_array(HashTable *regex, zval *replace, zend
 
 /* {{{ php_replace_in_subject
  */
-static zend_always_inline zend_string *php_replace_in_subject(zval *regex, zval *replace, zval *subject, size_t limit, size_t *replace_count)
+static zend_always_inline zend_string *php_replace_in_subject(
+	zend_string *regex_str, HashTable *regex_ht,
+	zend_string *replace_str, HashTable *replace_ht,
+	zend_string *subject, size_t limit, size_t *replace_count)
 {
 	zend_string *result;
-	zend_string *subject_str = zval_get_string(subject);
-
-	if (Z_TYPE_P(regex) != IS_ARRAY) {
-		result = php_pcre_replace(Z_STR_P(regex),
-								  subject_str,
-								  ZSTR_VAL(subject_str),
-								  ZSTR_LEN(subject_str),
-								  Z_STR_P(replace),
-								  limit,
-								  replace_count);
-		zend_string_release_ex(subject_str, 0);
+
+	if (regex_str) {
+		ZEND_ASSERT(replace_str != NULL);
+		result = php_pcre_replace(regex_str, subject, ZSTR_VAL(subject), ZSTR_LEN(subject),
+			replace_str, limit, replace_count);
 	} else {
-		result = php_pcre_replace_array(Z_ARRVAL_P(regex),
-										replace,
-										subject_str,
-										limit,
-										replace_count);
+		zend_string *subject_str = zend_string_copy(subject);
+		ZEND_ASSERT(regex_ht != NULL);
+		result = php_pcre_replace_array(regex_ht, replace_str, replace_ht, subject_str,
+			limit, replace_count);
 	}
 	return result;
 }
 /* }}} */
 
 /* {{{ php_replace_in_subject_func
  */
-static zend_string *php_replace_in_subject_func(zval *regex, zend_fcall_info *fci, zend_fcall_info_cache *fcc, zval *subject, size_t limit, size_t *replace_count, zend_long flags)
+static zend_string *php_replace_in_subject_func(zend_string *regex_str, HashTable *regex_ht,
+	zend_fcall_info *fci, zend_fcall_info_cache *fcc,
+	zend_string *subject, size_t limit, size_t *replace_count, zend_long flags)
 {
 	zend_string *result;
-	zend_string	*subject_str = zval_get_string(subject);
 
-	if (Z_TYPE_P(regex) != IS_ARRAY) {
+	if (regex_str) {
 		result = php_pcre_replace_func(
-			Z_STR_P(regex), subject_str, fci, fcc, limit, replace_count, flags);
-		zend_string_release_ex(subject_str, 0);
+			regex_str, subject, fci, fcc, limit, replace_count, flags);
 		return result;
 	} else {
+		/* If regex is an array */
 		zval		*regex_entry;
 
-		/* If regex is an array */
+		ZEND_ASSERT(regex_ht != NULL);
```

Comment:
Don't we need to addref the subject in this branch?

Reply:
Indeed